### PR TITLE
ContextCache usage cleanup

### LINKED ISSUE #519: Issue with ContextCache
See this issue
https://github.com/choderalab/perses/issues/613

And this comment:
https://github.com/choderalab/perses/issues/613#issuecomment-902512392

The current workaround is to create a instance of `ContextCache` and use that instead of the default global one, ie.

```python
context_cache = cache.ContextCache(capacity=None, time_to_live=None)
move = mcmc.LangevinSplittingDynamicsMove(
    timestep=4.0 * unit.femtoseconds,
    n_steps=250,
    collision_rate=5.0 / unit.picosecond,
    reassign_velocities=False,
    n_restart_attempts=20,
    constraint_tolerance=1e-06,
    context_cache=context_cache,  # This is the important part of the example
)
```

### LINKED ISSUE #536: Upon resuming a simulation from a .nc, new context caches are created for each move
This is a follow up to the fix here: https://github.com/choderalab/openmmtools/pull/524, which prevents the creation of a new context cache for each move.

The fix works fine for starting a simulation, but when I resume a simulation from a .nc file, I run into memory issues. 

This is because [when mcmc moves are read from the reporter](https://github.com/choderalab/openmmtools/blob/master/openmmtools/multistate/multistatereporter.py#L752-L768), each move is deserialized (`deserialize()` defined [here](https://github.com/choderalab/openmmtools/blob/456af23d91d12d570ec755674716ab0003b69e5b/openmmtools/utils.py#L652)).  Each `deserialize()` call involves a `__setState__()` call. For `BaseIntegratorMove`, that's defined [here](https://github.com/choderalab/openmmtools/blob/master/openmmtools/mcmc.py#L760-L767). In the `BaseIntegratorMove`'s  `__setState__()` , the context cache is then deserialized. However, in the `ContextCache`'s `__setState__()` (defined [here](https://github.com/choderalab/openmmtools/blob/master/openmmtools/cache.py#L472-L482)), a new context cache is created each time. This means that when deserializing the mcmc moves from the reporter, we are re-introducing the problem that the aforementioned PR solves (i.e. each move now has its own context cache).

Btw, you can check that each move has a separate context cache from this code:
```python
from openmmtools.multistate import MultiStateReporter
path = '4_apo.nc'
reporter = MultiStateReporter(path, 'r')
moves = reporter.read_mcmc_moves()
print(id(moves[0].context_cache))
print(id(moves[1].context_cache))
```

Attached are files to reproduce: 
[debug-resume-memory-issue.zip](https://github.com/choderalab/openmmtools/files/7738057/debug-resume-memory-issue.zip)

You can try resuming the simulation with this code:
```python
import argparse
from openmmtools.multistate import MultiStateReporter
from perses.samplers.multistate import HybridRepexSampler
import logging
import os

# Load repex simulation
_logger = logging.getLogger()
_logger.setLevel(logging.DEBUG)
reporter_file = "4_apo.nc"
reporter = MultiStateReporter(reporter_file, checkpoint_interval=10)
simulation = HybridRepexSampler.from_storage(reporter)

# Determine how many more iterations are needed
total_iterations = 5000
iterations =  total_iterations - simulation.iteration

# Resume simulation
simulation.extend(n_iterations=iterations)
```

### PR DESCRIPTION
**Pull request by ijpulidos**
## Description
Given that the `ContextCache` is a performance _trick_. It is better that it doesn't get serialized at all and that it is only used when the moves are applied or energies are computed. 

The `mcmc` module classes won't have a `context_cache` kwarg, instead the `apply` method for each will get fed the `context_cache` from the `multistatesampler` module objects.

This should tackle the memory issues when dealing with multiple replicas (fixes #536  ) and also work around the cycling issue (fixes #519 ). 

## Todos
- [x] Implement feature / fix bug
- [x] Add [tests](https://github.com/choderalab/openmmtools/tree/master/openmmtools/tests)
- [ ] Update [documentation](https://github.com/choderalab/openmmtools/tree/master/docs) as needed
- [ ] Update [changelog](https://github.com/choderalab/openmmtools/blob/master/docs/releasehistory.rst)Notable points that this PR has either accomplished or will accomplish.

## Status
- [x] Ready to go

### CODE REVIEW

**Comment by ijpulidos** [comment]
@jchodera According to our discussion on how to refactor this so we can come up with a cleaner use of the `ContextCache`. I realized that the way I understood this should be change breaks the API in at least two ways:

* Move objects won't have now a `context_cache` attribute, since this will be fed by the `MultiStateSampler`
* Calling the move method `.apply()` now requires a new argument for the `ContextCache` object. This one can be worked around by making it a keyword argument and specify a default behavior, but we would then have to do this in every move that implements the `apply` method.

Is it okay breaking the API in this way or maybe I missed something and we should be able to make these changes without breaking the API at all?

**Comment by jchodera** [comment]
After our discussion, here is what we could do:
1. In order to support backwards-compatibility`MCMCMove` would receive an initializer that would accept an optional `context_cache` that would issue a warning if specified:
```python
def __init__(self, context_cache=None):
    if context_cache is not None:
        warning('Specifying context_cache on initialization of an MCMCMove is deprecated; please pass to apply() instead')
```
2. All `MCMCMove` direct subclasses would call `super().__init__(**kwargs)` and accept `**kwargs` in order to continue to allow `context_cache` as an optional argument, which would issue this warning. No base class initializers would store or directly accept `context_cache` anymore
3. All `MCMCMove` implementations of `apply()` would now accept optional keyword argument `context_cache=None`, and then use an `MCMCMove` base implementation of `_get_context_cache()` to return which context cache to use:
```python
context_cache = self._get_context_cache(context_cache)
```

**Comment by ijpulidos** [comment]
With this approach I'm seeing a potential issue with the `mcmc_move.context_cache` attribute. We have to make sure we are syncing it and updating it correctly. Can we expect this to be done **exclusively** by invoking the `apply()` method?

As we have it in this moment we would need to sync this attribute in the `apply` method but we can get strange/undefined behavior when we just initialize the object without having run `apply` (the attribute is fixed as `None` in this case).

**Comment by ijpulidos** [comment]
From our dev sync, we will be returning the **global** context cache for the attribute (maintaining backwards compatibility), raising a warning since we expect users to use the `context_cache`  parameter of the `apply()` method, instead.

**Comment by ijpulidos** [comment]
This should be pretty close to be finished. Of course, the CI workflow test pass has the last word. Locally tests_mcmc.py are passing but we also need test_sampling to pass (that one takes a long time, so I'll just wait for CI to catch anything). Opening again for review.

**Comment by ijpulidos** [comment]
@jchodera @mikemhenry  Now that the tests are happy with the changes, can you take a look at it? I think this is ready to go, but it resulted in a somewhat bigger change than expected, so reviews are very appreciated :).

**Comment by ijpulidos** [comment]
I will be testing this further to see if the cycling issue is not shown (with the default behavior, #519 ) and that the context caches are shared when resuming simulations( #536 )

File: openmmtools/mcmc.py
Comment (jchodera) [inline openmmtools/mcmc.py:155]: Is this only to support subclasses of `MCMCMove` that may reference `self.context_cache`? 
If so, I wonder if we want to make this a `@property` so we can also issue the warning when someone tries to access `self.context_cache`?
```python
@property
def context_cache(self):
    logger.warning('The MCMCMove.context_cache field has been deprecated. The API now requires context_cache be passed to apply(). Please update your code.')
    from openmmtools import cache
    return cache.global_context_cache
```

File: openmmtools/mcmc.py
Comment (jchodera) [inline openmmtools/mcmc.py:157]: What is this used for? Do we need it?

File: openmmtools/mcmc.py
Comment (jchodera) [inline openmmtools/mcmc.py:175]: This is incomplete. It should be:
```
        context_cache_input : openmmtools.cache.ContextCache or None
            Context cache to be used in the propagation of the mcmc move. If None,
            it will create a new unlimited ContextCache object.
```

File: openmmtools/mcmc.py
Comment (jchodera) [inline openmmtools/mcmc.py:190]: If `None`, it should return the global ContextCache, not create a new one!

**Review by jchodera — COMMENTED** [review]
@ijpulidos : I think there may be some confusion regarding `MCMCMove._get_context_cache()`. 

When we discussed this, we wanted to remove any methods or fields specifying `MCMCMove.context_cache` and localize it _only_ in the `MCMCMove.apply()` method. 

However, the `MultistateSampler` needs a `_get_context_caches(propagation_cache=None, energy_cache=None)` method that will either use the specified context cache or the global context cache.

I'm not sure why we need an `MCMCMove._get_context_cache()`. Perhaps we can have a brief chat to sort this out?

File: openmmtools/multistate/multistatesampler.py
Comment (jchodera) [inline openmmtools/multistate/multistatesampler.py:1651]: We can just call the arguments `energy_context_cache` and `propagation_context_cache`. We don't need the `_input` suffix.

File: openmmtools/multistate/multistatesampler.py
Comment (jchodera) [inline openmmtools/multistate/multistatesampler.py:1688]: This method is internal, so it can just store these as `self.energy_context_cache` and `self.propagation_context_cache`. It doesn't need to return them.

File: openmmtools/multistate/sams.py
Comment (jchodera) [inline openmmtools/multistate/sams.py:366]: Should we just use `**kwargs` so any kwargs can be passed to the base class?

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:180]: From our discussion, we find this useful but we want to change the behavior to use a global context cache if None.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:157]: So, we needed this for the test [here](https://github.com/choderalab/openmmtools/blob/clean-context-cache-usage/openmmtools/tests/test_mcmc.py#L255-L265). I'm refactoring this test to use the number of accesses instead of having to use this private attribute that's only useful in testing. In the end this attribute won't be needed, thanks!

**Comment by mikemhenry** [comment]
How does this PR interact with:
https://github.com/choderalab/openmmtools/pull/537

**Comment by ijpulidos** [comment]
> How does this PR interact with: #537

@mikemhenry This PR should resolve the issue that #537 solves, so this one supersedes it. Thanks for noticing, I will update the description and close PRs accordingly, once this PR gets completed.

File: openmmtools/multistate/multistatesampler.py
Comment (ijpulidos) [reply openmmtools/multistate/multistatesampler.py:1688]: Here I was just thinking about explicitly setting the attributes (even if private/internal) in the `__init__`, because, as far as I know, it is commonly frowned upon to set them in another method, even if the method is called in the `__init__` I think it makes it a little cryptic. What do you think?

Comment (jchodera) [reply openmmtools/multistate/multistatesampler.py:1688]: That's fine---I agree your approach is less confusing.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py]: Should be fixed now in latest version.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py]: Now we don't need this attribute and testing is now using number of accesses.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py]: It's now using the global context cache by default.

File: openmmtools/multistate/multistatesampler.py
Comment (ijpulidos) [reply openmmtools/multistate/multistatesampler.py]: Fixed in latest changes.

File: openmmtools/multistate/sams.py
Comment (ijpulidos) [reply openmmtools/multistate/sams.py]: Using arbitrary keyword arguments (kwargs) now. Do we want to do the same for non-keyword arguments(`reporter`) here? Seems like it might make it less readable :thinking:

File: openmmtools/mcmc.py
Comment (jchodera) [reply openmmtools/mcmc.py]: Shouldn't we delete the attribute then?

**Comment by ijpulidos** [comment]
I think this is expected, but just so that we are clear. Since we **were** serializing some information (metadata) about the context caches before these changes, we are now not being able to deserialize `.nc` files that were serialized with previous versions of our software. I wonder how much of an issue this is. We could try doing something with the deserialization but it feels like it is just doing some monkey-patching and it's probably not worth the effort. Related to #536 

The good news is that if I use the script/files attached to this comment, we now get shared context caches upon resuming/deserialization of simulations (see again #536). I'm editing the description of this issue to match that.

**Comment by ijpulidos** [comment]
On other good news, the default behavior in this branch now avoids the cycling issue in #519 which is also related to https://github.com/choderalab/perses/issues/613 (where the main discussion is held). Also, it doesn't affect the performance when a context cache is specified. As far as I can see, then issue #519 is fixed (in terms of working around the cycling issue).

I am now waiting for the review on the new CI tests for the more up to date python versions (#542 ) which would also help the test passing on this PR.

**Comment by jchodera** [comment]
> I think this is expected, but just so that we are clear. Since we were serializing some information (metadata) about the context caches before these changes, we are now not being able to deserialize .nc files that were serialized with previous versions of our software. I wonder how much of an issue this is. We could try doing something with the deserialization but it feels like it is just doing some monkey-patching and it's probably not worth the effort. Related to #536

@ijpulidos : I think we had discussed that this could be handled by writing our own `MCMCMove` base class deserialization routine which would just silently drop the cache-related entries from the `dict` and then call `__setstate__` for the parent object. 

You're right that this may not be worth the effort, but it might mean we should not break backward-compatibility in a bigfix patch release, and should instead call this 0.10.0.

File: openmmtools/mcmc.py
Comment (jchodera) [inline openmmtools/mcmc.py:294]: The `MCMCSampler` isn't actually used in `MultiStateSampler`, and doesn't need separate `_energy_context_cache` or `_propagation_context_cache` attributes. 

In fact, we don't need to store any context caches at all---we can just accept an optional `context_cache` argument in `run()`, where we would default to `cache.global_context_cache` if none is specified.

File: openmmtools/mcmc.py
Comment (jchodera) [inline openmmtools/mcmc.py:358]: We wouldn't need this method either, since we would just either use the provided `context_cache` in `run()` or fall back to the `cache.global_context_cache`.

**Review by jchodera — CHANGES_REQUESTED** [review]
Thanks for the hard work on this PR, @ijpulidos!

For `MCMCSampler`: This is not actually used by `MultiStateSampler`---it's just an independent driver for sampling with `MCMCMove` that folks could use if they just want to simulate from a single thermodynamic state. As such, we don't need it to store any `ContextCache` objects internally---we can just ask the user to specify an optional `context_cache` when they call `run()` or fall back to the `cache.global_context_cache` otherwise.

For `MultiStateSampler`, I discovered an important issue: We don't have a way to ensure that the `ContextCache` we automatically create has the platform properties we need (e.g. `mixed` precision for CUDA or OpenCL; `DeterminsticForces` set to `true` for CUDA) if we need to create our own `ContextCache` objects, and there's nothing in `perses` that handles this either. 

Here's my suggested solution: Let's only accept `propagation_context_cache` and `energy_context_cache` arguments in `MultiStateSampler.run()`, and strip out all the other infrastructure that stores these as attributes. That will vastly simplify the infrastructure we need here.

`run()` would then fall back to `cache.global_context_cache` for any caches that are left as the default (`None`). This will cause periodic cycling if nothing else is done, but that's OK---we will create and configure these `ContextCache` arguments in perses explicitly and pass them in when we call `run()`.

We can also clean up some issues with `ContextCache` while we're at it:

We should fix [these lines](https://github.com/choderalab/openmmtools/blob/master/openmmtools/cache.py#L678-L680) in `ContextCache` to make sure they clean up:
```
        try:
            openmm.Context(dummy_system, dummy_integrator, platform, platform_properties)
            return True
```
by adding an explicit `del context`:
```
        try:
            context = openmm.Context(dummy_system, dummy_integrator, platform, platform_properties)
            del context
            return True
```

I also note that the `global_context_cache` is [created without any platform properties specified](https://github.com/choderalab/openmmtools/blob/master/openmmtools/cache.py#L780), so defaults to `single` precision on the fastest platform (e.g. CUDA if available). If we can figure out how to at least change this to `mixed` precision and set `DeterministicForces` to `true` by default, that would be ideal, but we may need to save that for a separate PR in the future and just write an issue to capture that future action item.

To create and configure the `ContextCache` objects in perses, we'll want to do something like
```python
# Attempt to get Platform
# NOTE: See openmmtools.utils.get_fastest_platform() if we just want the fastest platform that supports mixed precision
platform = openmm.Platform.getPlatformByName(platform_name)
# Configure platofrm
if platform.getName() in ['CUDA', 'OpenCL']:
    platform.setPropertyDefaultValue('Precision', 'mixed')
if platform.getName() == 'CUDA':
    platform.setPropertyDefaultValue('DeterministicForces', 'true')
# Create ContextCache for this platform
context_cache = ContextCache(platform=platform)
```

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:210]: You are right, now that the `context_cache` property defaults to the global context cache we shouldn't be needing this,

**Comment by ijpulidos** [comment]
@jchodera respect the deserialization issue, you are right we did talk about this and I did make this work. I was just having a different deserialization issue (because I was using the wrong branch for the test, this one needed the RESTCapable branch of perses) and thought it was due to my changes. So, serialization/deserialization should be working just fine. Sorry for the noise.

**Comment by mikemhenry** [comment]
@ijpulidos give me a ping on slack when this is ready for review

**Comment by ijpulidos** [comment]
@jchodera Your suggestions sound reasonable, this will reduce the complexity of the infrastructure when dealing with `ContextCache` objects. 

I think we still need the `MultiStateSampler.from_storage` method to accept both `energy_context_cache` and `propagation_context_cache` (so it would be both `.run` and `.from_storage` that will accept these). And we have to make users specify them when they try to resume a simulation (or load one from serialized objects) IF they don't want to hit the cycling issue. Did I get that correctly?

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:294]: As discussed we now have `energy_context_cache` and `sampler_context_cache` as attributes that are passed in `create()` and default to using the global context cache if `None`.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:358]: As discussed, this method is useful to have a centralized point where to handle the context caches behavior. This method is accesed from `create()` and `from_storage()`, so it reduces redundant code.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:294]: Ok, I'll make the changes, this one is still missing in the current status of the PR.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:210]: We still need it, because `MultiStateSampler` makes a deepcopy of the `mcmc_moves`, so we don't want each copy to have its own `context_cache` because it will end up eating memory without need (issue in #521 )

File: openmmtools/mcmc.py
Comment (jchodera) [reply openmmtools/mcmc.py:358]: OK!

File: openmmtools/mcmc.py
Comment (jchodera) [reply openmmtools/mcmc.py:210]: Can you elaborate a bit more?`MCMCMove` and derived classes no longer contain a `context_cache` attribute.

Comment (ijpulidos) [reply openmmtools/mcmc.py:210]: Oh, you are correct! it's not a property and this one shouldn't be affected by the deepcopying. I agree, I don't think we need it.

Comment (ijpulidos) [reply openmmtools/mcmc.py]: Changed in current state of the PR.

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py]: Changed in current state of the PR.

File: openmmtools/multistate/multistatesampler.py
Comment (jchodera) [inline openmmtools/multistate/multistatesampler.py:983]: I think my only comment is that we discussed this could be simplified into a single `self._initialize_context_cache()` that just returns the global context cache, and there is no need to accept the kwargs for context caches since we can just set them as public attributes now, but I don't feel very strongly about this either way. I think it's just simpler and more streamlined if we don't have to pass these through various ways of creating or initializing the class, but it's not too critical either way.

**Review by jchodera — APPROVED** [review]
This is looking good now, and it looks like everything passes!
It would be good to do some quick testing on lilac to make sure we have fixed periodic cycling when initialized appropriately from perses, and have @mikemhenry give it a look over before merging and cutting the new openmmtools release!

File: openmmtools/multistate/multistatesampler.py
Comment (ijpulidos) [reply openmmtools/multistate/multistatesampler.py:983]: I think I understand what you mean. For the sake of this release I think we are good with having this as it is now, but I'll raise a new issue where we can discuss enhancing this behavior.

**Comment by ijpulidos** [comment]
Cycling issue is not shown when initialized properly. @mikemhenry can you please review, thanks!

File: openmmtools/mcmc.py
Comment (mikemhenry) [inline openmmtools/mcmc.py:697]: Is this a future TODO or did you figure it out?

File: openmmtools/multistate/multistatesampler.py
Comment (mikemhenry) [reply openmmtools/multistate/multistatesampler.py:983]: Did you raise the issue? If you did can you link to John's comment?

**Review by mikemhenry — APPROVED** [review]
Just a few comments

File: openmmtools/mcmc.py
Comment (ijpulidos) [reply openmmtools/mcmc.py:697]: It's still to be figured out, it just seems a bit redundant to specify a context cache and then use this `get_context`, but we have you to find out if there's a better way to accomplish this. Maybe for a future quick enhancement?

**Pull request merged**
